In [1]:
import numpy as np
import pandas as pd

In [30]:
book = pd.read_csv("Books.csv")
user = pd.read_csv("Users.csv")
ratings = pd.read_csv("Ratings.csv")

C:\Users\welcome\AppData\Local\Temp\ipykernel_5744\2268120439.py:1: DtypeWarning: Columns (3) have mixed types. Specify dtype option on import or set low_memory=False.
  book = pd.read_csv("Books.csv")


In [5]:
book.head()

ISBN                                         Book-Title  \
0  0195153448                                Classical Mythology   
1  0002005018                                       Clara Callan   
2  0060973129                               Decision in Normandy   
3  0374157065  Flu: The Story of the Great Influenza Pandemic...   
4  0393045218                             The Mummies of Urumchi   

            Book-Author Year-Of-Publication                   Publisher  \
0    Mark P. O. Morford                2002     Oxford University Press   
1  Richard Bruce Wright                2001       HarperFlamingo Canada   
2          Carlo D'Este                1991             HarperPerennial   
3      Gina Bari Kolata                1999        Farrar Straus Giroux   
4       E. J. W. Barber                1999  W. W. Norton &amp; Company   

                                         Image-URL-S  \
0  http://images.amazon.com/images/P/0195153448.0...   
1  http://images.amazon.com/images/P/0002005018.0...   
2  http://images.amazon.com/images/P/0060973129.0...   
3  http://images.amazon.com/images/P/0374157065.0...   
4  http://images.amazon.com/images/P/0393045218.0...   

                                         Image-URL-M  \
0  http://images.amazon.com/images/P/0195153448.0...   
1  http://images.amazon.com/images/P/0002005018.0...   
2  http://images.amazon.com/images/P/0060973129.0...   
3  http://images.amazon.com/images/P/0374157065.0...   
4  http://images.amazon.com/images/P/0393045218.0...   

                                         Image-URL-L  
0  http://images.amazon.com/images/P/0195153448.0...  
1  http://images.amazon.com/images/P/0002005018.0...  
2  http://images.amazon.com/images/P/0060973129.0...  
3  http://images.amazon.com/images/P/0374157065.0...  
4  http://images.amazon.com/images/P/0393045218.0...

Try to know the dimensions of each of the three datasets

In [6]:
print(book.shape)
print(user.shape)
print(reviews.shape)

(271360, 8)
(278858, 3)
(1149780, 3)


finding amount of null data or missing data from data set

In [8]:
book.isnull().sum()

ISBN                   0
Book-Title             0
Book-Author            2
Year-Of-Publication    0
Publisher              2
Image-URL-S            0
Image-URL-M            0
Image-URL-L            3
dtype: int64

In [9]:
user.isnull().sum()

User-ID          0
Location         0
Age         110762
dtype: int64

age ka ham bohot zyada use nahi karenge so peace hai

In [10]:
ratings.isnull().sum()

User-ID        0
ISBN           0
Book-Rating    0
dtype: int64

finding duplicate values

In [14]:
book.duplicated().sum()
user.duplicated().sum()
ratings.duplicated().sum()

0

so basically no duplicates what so ever sexy

we are gonna do pandas profiling 

we did it and have three html files for the same

In [ ]:
pip install --upgrade jupyterlab

In [ ]:

pip install pandas-profiling

now we start doing model deployment first e are making top 50 popularity based reccomendation just like the trending tab of youtube

#  Popularity based system

In [33]:
rating_with_name = pd.merge(ratings,book,how='inner',on='ISBN')

In [ ]:
rating_with_name.sample(5)

In [37]:
rating_with_name.groupby('Book-Title').count()

User-ID  ISBN  \
Book-Title                                                          
 A Light in the Storm: The Civil War Diary of A...        4     4   
 Always Have Popsicles                                    1     1   
 Apple Magic (The Collector's series)                     1     1   
 Ask Lily (Young Women of Faith: Lily Series, B...        1     1   
 Beyond IBM: Leadership Marketing and Finance f...        1     1   
...                                                     ...   ...   
Ã?Â?lpiraten.                                             2     2   
Ã?Â?rger mit Produkt X. Roman.                            4     4   
Ã?Â?sterlich leben.                                       1     1   
Ã?Â?stlich der Berge.                                     3     3   
Ã?Â?thique en toc                                         2     2   

                                                    Book-Rating  Book-Author  \
Book-Title                                                                     
 A Light in the Storm: The Civil War Diary of A...            4            4   
 Always Have Popsicles                                        1            1   
 Apple Magic (The Collector's series)                         1            1   
 Ask Lily (Young Women of Faith: Lily Series, B...            1            1   
 Beyond IBM: Leadership Marketing and Finance f...            1            1   
...                                                         ...          ...   
Ã?Â?lpiraten.                                                 2            2   
Ã?Â?rger mit Produkt X. Roman.                                4            4   
Ã?Â?sterlich leben.                                           1            1   
Ã?Â?stlich der Berge.                                         3            3   
Ã?Â?thique en toc                                             2            2   

                                                    Year-Of-Publication  \
Book-Title                                                                
 A Light in the Storm: The Civil War Diary of A...                    4   
 Always Have Popsicles                                                1   
 Apple Magic (The Collector's series)                                 1   
 Ask Lily (Young Women of Faith: Lily Series, B...                    1   
 Beyond IBM: Leadership Marketing and Finance f...                    1   
...                                                                 ...   
Ã?Â?lpiraten.                                                         2   
Ã?Â?rger mit Produkt X. Roman.                                        4   
Ã?Â?sterlich leben.                                                   1   
Ã?Â?stlich der Berge.                                                 3   
Ã?Â?thique en toc                                                     2   

                                                    Publisher  Image-URL-S  \
Book-Title                                                                   
 A Light in the Storm: The Civil War Diary of A...          4            4   
 Always Have Popsicles                                      1            1   
 Apple Magic (The Collector's series)                       1            1   
 Ask Lily (Young Women of Faith: Lily Series, B...          1            1   
 Beyond IBM: Leadership Marketing and Finance f...          1            1   
...                                                       ...          ...   
Ã?Â?lpiraten.                                               2            2   
Ã?Â?rger mit Produkt X. Roman.                              4            4   
Ã?Â?sterlich leben.                                         1            1   
Ã?Â?stlich der Berge.                                       3            3   
Ã?Â?thique en toc                                           2            2   

                                                    Image-URL-M  Image-URL-L  
Book-Title                                 

In [40]:
rating_with_name.groupby('Book-Title').count()['Book-Rating'].reset_index()

Book-Title  Book-Rating
0        A Light in the Storm: The Civil War Diary of ...            4
1                                   Always Have Popsicles            1
2                    Apple Magic (The Collector's series)            1
3        Ask Lily (Young Women of Faith: Lily Series, ...            1
4        Beyond IBM: Leadership Marketing and Finance ...            1
...                                                   ...          ...
241066                                      Ã?Â?lpiraten.            2
241067                     Ã?Â?rger mit Produkt X. Roman.            4
241068                                Ã?Â?sterlich leben.            1
241069                              Ã?Â?stlich der Berge.            3
241070                                  Ã?Â?thique en toc            2

[241071 rows x 2 columns]

yahan basically groupby kiya and usse counts aaye yahan counts matlab is book ki entries mein kitne baar aaya vo data frame mein matlab
group by min jo rating hai na vo ye nahi ki kitne stars hain balki  ye hai ki kitne logon yaani kitne users ne rate kiya hai isi book ko

[['Book-Rating'] ka matlab ki ye jo df aayega naya group by karne ke baad usme kewal Book rating ka coloumn do along with title 
now groupby ke karan title hi index ban gaya hai isiliye  they want ki wapas se simple 1,2 ,3 ,4 index aajaye isiliye reset index

In [61]:
num_rating=rating_with_name.groupby('Book-Title').count()['Book-Rating'].reset_index()

yahan bas wahi rename kiya taaki  apne ko confusion na ho

In [78]:
num_rating.rename(columns={'Book-Rating':'number_of_ratings'},inplace=True)

yahan basically ek naya data frame banya jisme sirf books title and unki mean ratings hai is baar count nahi ki kitni rating hai bas avergae rating

In [79]:
average_ratings = rating_with_name.groupby('Book-Title')['Book-Rating'].mean().reset_index()

In [80]:
average_ratings.rename(columns={'Book-Rating':'average_ratings'},inplace=True)

now we will merge avegrating df and nummber of rating df as usse hame ye milega book title,number of ratings,average rating 

In [81]:
rating_full_info = pd.merge(average_ratings,num_rating,how="inner",on="Book-Title").reset_index()

now we ant books with highest number of ratings as they are the most reviewed and usme se highest avergae rating wali 100 nikal lenge as we want those ith highest number of reviews and good reviews

In [82]:
rating_full_info[rating_full_info['number_of_ratings']>255]

index                                         Book-Title  \
764        764                                               1984   
818        818                                1st to Die: A Novel   
1048      1048                                         2nd Chance   
1760      1760                                 A Bend in the Road   
2281      2281  A Child Called \It\": One Child's Courage to S...   
...        ...                                                ...   
233850  233850                           White Oleander : A Novel   
233851  233851       White Oleander : A Novel (Oprah's Book Club)   
234740  234740  Wicked: The Life and Times of the Wicked Witch...   
234951  234951                                        Wild Animus   
240860  240860                                \O\" Is for Outlaw"   

        average_ratings  number_of_ratings  
764            4.454225                284  
818            3.575639                509  
1048           3.269663                356  
1760           3.364162                346  
2281           4.086792                265  
...                 ...                ...  
233850         3.506460                387  
233851         3.772472                356  
234740         3.766871                326  
234951         1.019584               2502  
240860         2.988848                269  

[182 rows x 4 columns]

In [84]:
rating_full_info[rating_full_info['number_of_ratings']>255].sort_values('average_ratings', ascending=False)

index                                         Book-Title  \
80434    80434  Harry Potter and the Prisoner of Azkaban (Book 3)   
80422    80422       Harry Potter and the Goblet of Fire (Book 4)   
80441    80441     Harry Potter and the Sorcerer's Stone (Book 1)   
80426    80426  Harry Potter and the Order of the Phoenix (Boo...   
80414    80414   Harry Potter and the Chamber of Secrets (Book 2)   
...        ...                                                ...   
227701  227701       Vinegar Hill (Oprah's Book Club (Paperback))   
233635  233635                                           Whispers   
143377  143377                                  Presumed Innocent   
94382    94382                                       Isle of Dogs   
234951  234951                                        Wild Animus   

        average_ratings  number_of_ratings  
80434          5.852804                428  
80422          5.824289                387  
80441          5.737410                278  
80426          5.501441                347  
80414          5.183453                556  
...                 ...                ...  
227701         2.245283                265  
233635         2.199301                286  
143377         2.139456                294  
94382          2.000000                288  
234951         1.019584               2502  

[182 rows x 4 columns]

till now we got book titles with more than 255 number of ratings and they are sorted according to thier average rating in descending order

now inme se top 100 nikalenge apan

In [85]:
top_100_books= rating_full_info[rating_full_info['number_of_ratings']>255].sort_values('average_ratings', ascending=False).head(100)

In [86]:
top_100_books.head()

index                                         Book-Title  \
80434  80434  Harry Potter and the Prisoner of Azkaban (Book 3)   
80422  80422       Harry Potter and the Goblet of Fire (Book 4)   
80441  80441     Harry Potter and the Sorcerer's Stone (Book 1)   
80426  80426  Harry Potter and the Order of the Phoenix (Boo...   
80414  80414   Harry Potter and the Chamber of Secrets (Book 2)   

       average_ratings  number_of_ratings  
80434         5.852804                428  
80422         5.824289                387  
80441         5.737410                278  
80426         5.501441                347  
80414         5.183453                556

In [88]:
pd.merge(top_100_books,book,how="inner",on="Book-Title")

index                                         Book-Title  \
0     80434  Harry Potter and the Prisoner of Azkaban (Book 3)   
1     80434  Harry Potter and the Prisoner of Azkaban (Book 3)   
2     80434  Harry Potter and the Prisoner of Azkaban (Book 3)   
3     80422       Harry Potter and the Goblet of Fire (Book 4)   
4     80422       Harry Potter and the Goblet of Fire (Book 4)   
..      ...                                                ...   
348   93302  Into Thin Air : A Personal Account of the Mt. ...   
349  201472                                        The Partner   
350  201472                                        The Partner   
351  201472                                        The Partner   
352  201472                                        The Partner   

     average_ratings  number_of_ratings        ISBN    Book-Author  \
0           5.852804                428  0439136350  J. K. Rowling   
1           5.852804                428  0439136369  J. K. Rowling   
2           5.852804                428  0786222743  J. K. Rowling   
3           5.824289                387  0439139597  J. K. Rowling   
4           5.824289                387  0439139600  J. K. Rowling   
..               ...                ...         ...            ...   
348         3.303763                372  0385492081   JON KRAKAUER   
349         3.290795                478  0385472951   John Grisham   
350         3.290795                478  0440224764   John Grisham   
351         3.290795                478  0440295556   John Grisham   
352         3.290795                478  0553472836   John Grisham   

    Year-Of-Publication                Publisher  \
0                  1999               Scholastic   
1                  2001               Scholastic   
2                  2000          Thorndike Press   
3                  2000               Scholastic   
4                  2002    Scholastic Paperbacks   
..                  ...                      ...   
348                1998                   Anchor   
349                1997          Doubleday Books   
350                1998  Dell Publishing Company   
351                1997  Dell Publishing Company   
352                1997       Random House Audio   

                                           Image-URL-S  \
0    http://images.amazon.com/images/P/0439136350.0...   
1    http://images.amazon.com/images/P/0439136369.0...   
2    http://images.amazon.com/images/P/0786222743.0...   
3    http://images.amazon.com/images/P/0439139597.0...   
4    http://images.amazon.com/images/P/0439139600.0...   
..                                                 ...   
348  http://images.amazon.com/images/P/0385492081.0...   
349  http://images.amazon.com/images/P/0385472951.0...   
350  http://images.amazon.com/images/P/0440224764.0...   
351  http://images.amazon.com/images/P/0440295556.0...   
352  http://images.amazon.com/images/P/0553472836.0...   

                                           Image-URL-M  \
0    http://images.amazon.com/images/P/0439136350.0...   
1    http://images.amazon.com/images/P/0439136369.0...   
2    http://images.amazon.com/images/P/0786222743.0...   
3    http://images.amazon.com/images/P/0439139597.0...   
4    http://images.amazon.com/images/P/0439139600.0...   
..                                                 ...   
348  http://images.amazon.com/images/P/0385492081.0...   
349  http://images.amazon.com/images/P/0385472951.0...   
350  http://images.amazon.com/images/P/0440224764.0...   
351  http://images.amazon.com/images/P/0440295556.0...   
352  http://images.amazon.com/images/P/0553472836.0...   

                                           Image-URL-L  
0    http://images.amazon.com/images/P/0439136350.0...  
1    http://images.amazon.com/images/P/0439136369.0...  
2    http://images.amazon.com/images/P/0786222743.0...  
3    http://images.amazon.com/images/P/0439139597.0...  
4    http://images.amazon.com/images/P/0439139600.0...  
..         

In [89]:
top_100_books = pd.merge(top_100_books,book,how="inner",on="Book-Title")

In [90]:
top_100_books.shape

(353, 11)

now why do we have 353 rows??

it's becuase same tittle of books are reviewed and have diffrent ISBN with them
now we will remove duplicate based on book title i.e if the book title is same then we will remove all but one as sabme same avergae rating and number of rarings hai

In [92]:
top_100_books.drop_duplicates("Book-Title")

index                                         Book-Title  \
0     80434  Harry Potter and the Prisoner of Azkaban (Book 3)   
3     80422       Harry Potter and the Goblet of Fire (Book 4)   
5     80441     Harry Potter and the Sorcerer's Stone (Book 1)   
9     80426  Harry Potter and the Order of the Phoenix (Boo...   
13    80414   Harry Potter and the Chamber of Secrets (Book 2)   
..      ...                                                ...   
341   89780                                         Icy Sparks   
343   14351                    Angela's Ashes (MMP) : A Memoir   
344   86511                                     How to Be Good   
347   93302  Into Thin Air : A Personal Account of the Mt. ...   
349  201472                                        The Partner   

     average_ratings  number_of_ratings        ISBN       Book-Author  \
0           5.852804                428  0439136350     J. K. Rowling   
3           5.824289                387  0439139597     J. K. Rowling   
5           5.737410                278  0590353403     J. K. Rowling   
9           5.501441                347  043935806X     J. K. Rowling   
13          5.183453                556  0439064872     J. K. Rowling   
..               ...                ...         ...               ...   
341         3.346278                309  0142000205  Gwyn Hyman Rubio   
343         3.337423                326  0684872153     Frank McCourt   
344         3.305921                304  1573229326       Nick Hornby   
347         3.303763                372  0385494785      JON KRAKAUER   
349         3.290795                478  0385472951      John Grisham   

    Year-Of-Publication        Publisher  \
0                  1999       Scholastic   
3                  2000       Scholastic   
5                  1998       Scholastic   
9                  2003       Scholastic   
13                 2000       Scholastic   
..                  ...              ...   
341                2001    Penguin Books   
343                1999         Scribner   
344                2002  Riverhead Books   
347                1999           Anchor   
349                1997  Doubleday Books   

                                           Image-URL-S  \
0    http://images.amazon.com/images/P/0439136350.0...   
3    http://images.amazon.com/images/P/0439139597.0...   
5    http://images.amazon.com/images/P/0590353403.0...   
9    http://images.amazon.com/images/P/043935806X.0...   
13   http://images.amazon.com/images/P/0439064872.0...   
..                                                 ...   
341  http://images.amazon.com/images/P/0142000205.0...   
343  http://images.amazon.com/images/P/0684872153.0...   
344  http://images.amazon.com/images/P/1573229326.0...   
347  http://images.amazon.com/images/P/0385494785.0...   
349  http://images.amazon.com/images/P/0385472951.0...   

                                           Image-URL-M  \
0    http://images.amazon.com/images/P/0439136350.0...   
3    http://images.amazon.com/images/P/0439139597.0...   
5    http://images.amazon.com/images/P/0590353403.0...   
9    http://images.amazon.com/images/P/043935806X.0...   
13   http://images.amazon.com/images/P/0439064872.0...   
..                                                 ...   
341  http://images.amazon.com/images/P/0142000205.0...   
343  http://images.amazon.com/images/P/0684872153.0...   
344  http://images.amazon.com/images/P/1573229326.0...   
347  http://images.amazon.com/images/P/0385494785.0...   
349  http://images.amazon.com/images/P/0385472951.0...   

                                           Image-URL-L  
0    http://images.amazon.com/images/P/0439136350.0...  
3    http://images.amazon.com/images/P/0439139597.0...  
5    http://images.amazon.com/images/P/0590353403.0...  
9    http://images.amazon.com/images/P/043935806X.0...  
13   http://images.amazon.com/images/P/0439064872.0...  
..                                                 ...  
341  http://im

In [102]:
top_100_book = top_100_book.drop_duplicates("Book-Title")

now see we only have 100 ros noice!!!!!!!!!!

now we will eliminate non use ful coloumns like we want one image url to siplay image ,we want book title,authore name number of rating and average reviews

In [103]:
top_100_book=top_100_book[["Book-Title","average_ratings","number_of_ratings","Book-Author","Year-Of-Publication","Image-URL-M"]]

In [105]:
top_100_book.shape

(100, 6)

In [106]:
top_100_book.to_csv('top_100_book.csv', index=False)